forum/src/main/java/unipd/se18/ocrcamera/forum/viewmodels/Login_VM.java

```diff
@@ -0,0 +1,111 @@
+
+import unipd.se18.ocrcamera.forum.DatabaseManager;
+import unipd.se18.ocrcamera.forum.R;
+
+/**
+ * View model that contains all the logic needed to perform a login by querying the database
+ * @author Leonardo Rossi (g2), Alberto Valente (g2), Taulant Bullaku (g2)
+ */
+public class Login_VM extends ViewModel implements LoginMethods {
+
+    /**
+     * ********************
+     * **   LISTENERS    **
+     * ********************
+     */
+
+    /**
+     * Listener useful for communicating with the View
+     * @see <a href="https://docs.oracle.com/javase/tutorial/uiswing/events/index.html">
+     *     Writing Event Listeners</a>
+     * @author Taulant Bullaku (g2)
+     */
+    public interface ForumLoginListener {
+
+        /**
+         * Triggered when the login request ends up successfully
+         * @param username The username of the user that successfully logged in
+         */
+        void onLoginSuccess(String username);
+
+        /**
+         * Triggered when the login request fails
+         * @param message The error message about what was wrong with the login request
+         */
+        void onLoginFailure(String message);
+
+    }
+
+    /**
+     * ***************************
+     * **   GLOBAL VARIABLES    **
+     * ***************************
+     */
+    private ForumLoginListener forumLoginListener;
+
+    /**
+     * String used for debug logs to identify the view model throwing it
+     */
+    private final String LOG_TAG = "@@Login_VM";
+
+    /**
+     * String used to identify the error given by incorrect credentials
+     */
+    private final String LOG_INCORRECT_CREDENTIALS = "Connection established. Credentials refused.";
+
+    @Override
+    public void loginToForum(final Context context, final String username, String password) {
+
+        /**
+         * Initialization of the listener useful to react to responses
+         * from the database requests, as soon as they are available
+         */
+        final DatabaseManager.Listeners listeners = new DatabaseManager.Listeners();
+
+        //Definition of the listener that will be triggered when the login process finishes
+        listeners.completeListener = new OnCompleteListener<QuerySnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<QuerySnapshot> task)
+            {
+                if (task.isSuccessful())
+                {
+                    if (task.getResult().size() == 1)
+                    {
+                        forumLoginListener.onLoginSuccess(username);
+                    }
+                    else
+                    {
+                        forumLoginListener.onLoginFailure(context.getString(R.string.loginFailedMessage));
+                    }
+                }
+            }
+        };
+
+        //Sends a netword request to check if the provided credentials are correct
+        DatabaseManager.loginUser(context, username, password, listeners);
+    }
```

Comment:
I see there are a lot of listeners, i suggest you to use different names to differentiate them. I had some difficulties to recognize them.

Reply:
You're right: I have replaced them with "forumLoginListeners" instead.